internal/common/auth/authorization/kubernetes.go

```diff
@@ -0,0 +1,203 @@
+package authorization
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/grpc-ecosystem/go-grpc-middleware/util/metautils"
+	"github.com/patrickmn/go-cache"
+	authv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/G-Research/armada/internal/common/auth/configuration"
+)
+
+type KubernetesNativeAuthService struct {
+	KidMappingFileLocation string
+	TokenCache             *cache.Cache
+	InvalidTokenExpiry     int64
+}
+
+func NewKubernetesNativeAuthService(config configuration.KubernetesAuthConfig) KubernetesNativeAuthService {
+	cache := cache.New(5*time.Minute, 5*time.Minute)
+	return KubernetesNativeAuthService{
+		KidMappingFileLocation: config.KidMappingFileLocation,
+		TokenCache:             cache,
+		InvalidTokenExpiry:     config.InvalidTokenExpiry,
+	}
+}
+
+type CacheData struct {
+	Name  string `json:"name"`
+	Valid bool   `json:"valid"`
+}
+
+func (authService *KubernetesNativeAuthService) Authenticate(ctx context.Context) (Principal, error) {
+	// Retrieve token from context.
+	authHeader := strings.SplitN(metautils.ExtractIncoming(ctx).Get("authorization"), " ", 2)
+
+	if len(authHeader) < 2 || authHeader[0] != "KubernetesAuth" {
+		return nil, missingCredentials
+	}
+
+	token, ca, err := parseAuth(authHeader[1])
+	if err != nil {
+		return nil, missingCredentials
+	}
+
+	// Get token time
+	expirationTime, err := parseTime(token)
+	if err != nil {
+		return nil, err
+	}
+
+	if time.Now().After(*expirationTime) {
```

Comment:
I've put a comment lower down to say that parseTime should return a struct rather than a pointer, but if you decide not to do that then I'd add a nil check here.